core/modules.rs

```diff
@@ -648,17 +603,32 @@ impl ModuleMap {
     self.info.get(id)
   }
 
-  pub fn load_main(
+  pub async fn load_main(
     &self,
     specifier: &str,
     code: Option<String>,
-  ) -> RecursiveModuleLoad {
-    RecursiveModuleLoad::main(
+  ) -> Result<RecursiveModuleLoad, AnyError> {
+    let module_specifier =
+      self
+        .loader
+        .resolve(self.op_state.clone(), specifier, ".", true)?;
+    let load = RecursiveModuleLoad::main(
```

Comment:
I don't really understand changes in this file. Please explain.

Reply:
Other than adding [this code path](https://github.com/denoland/deno/pull/11015/files#diff-aa01489135c068a2056422f99878ba91861c585baed075470c08c1ac439ab3b2R654-R655) to short-circuit preparing dynamic imported modules already in the module map, there are two refactors:
1. Move [this code](https://github.com/denoland/deno/pull/11015/files#diff-5b779288098bb6449725245ce7fdb1ec3190563b5dd255e747cfc35e639407aeL1258) (after the singular call site of `ModuleMap::load_main()`) inside of `ModuleMap::load_main()` to make its functionality more similar to `ModuleMap::load_dynamic_import()`.
2. Remove and inline `RecursiveModuleLoad::prepare()` at its two call sites in `ModuleMap::load_main()` and `ModuleMap::load_dynamic_import()`. Something like `RecursiveModuleLoad::prepare()` should probably still exist, but that would lead to a duplicate resolve in `ModuleMap::load_dynamic_import()`.

Reply:
> Other than adding this code path to short-circuit preparing dynamic imported modules already in the module map, there are two refactors:


I'm a bit weary about it. This function is one of problems in #8421 and I'm making strides to rewrite it into a future instead.

> - Move this code (after the singular call site of ModuleMap::load_main()) inside of ModuleMap::load_main() to make its functionality more similar to ModuleMap::load_dynamic_import(). 
> - Remove and inline RecursiveModuleLoad::prepare() at its two call sites in ModuleMap::load_main() and ModuleMap::load_dynamic_import(). Something like RecursiveModuleLoad::prepare() should probably still exist, but that would lead to a duplicate resolve in ModuleMap::load_dynamic_import().

So these two are just cleanups to make more of the APIs internal to `ModuleMap`. Can't `RecursiveModuleLoad::prepare()` be left as is and called by those two related methods? I'm worried they might diverge at some point.

Anyway, the tests look good and are passing so I think I'll land it and worry about potential troubles with `is_registered` later

Reply:
Though I'd like to get @kitsonk's review as well as this PR touches the module graph.

Reply:
> Can't `RecursiveModuleLoad::prepare()` be left as is and called by those two related methods? I'm worried they might diverge at some point.

Diverging isn't a problem, the two call sites after still share nothing except for a `prepare_load()` call. You'll notice `RecursiveModuleLoad::prepare()` had two separate code paths for if it was called from `load_main()` or `load_dynamic_import()`, which is why I thought to inline it. The amount of shared code is the same as it was before.

It's awkward because `RecursiveModuleLoad` would ideally keep a method for preparing itself, but as I mentioned that creates complications as to where and how many times `resolve()` is called in `load_dynamic_import()`.

Reply:
> Diverging isn't a problem, the two call sites after still share nothing except for a prepare_load() call. You'll notice RecursiveModuleLoad::prepare() had two separate code paths for if it was called from load_main() or load_dynamic_import(), which is why I thought to inline it. The amount of shared code is the same as it was before.

Of course 🤦‍♂️ that makes perfect sense

It's awkward because RecursiveModuleLoad would ideally keep a method for preparing itself, but as I mentioned that creates complications as to where and how many times resolve() is called in load_dynamic_import().

Is it a problem that resolve is called multiple times? It should be idempotent after all.

Reply:
Okay, I've brought back `RecursiveModuleLoad::prepare()`. I think `resolve()` has duped calls in other places anyway. On another note I want to see if I can make simplifications to `RecursiveModuleLoad::state`/`LoadState`.

Reply:
@nayeemrmn just a heads up that I'm also working on `core/modules.rs` in #10809

Reply:
I've replaced the following structure:
```rust
struct RecursiveModuleLoad {
  state: LoadState,
  // ...
}

enum LoadState {
  ResolveMain(String, Option<String>),
  ResolveImport(String, String),
  LoadingRoot,
  LoadingImports,
  Done,
}
```
with:
```rust
struct RecursiveModuleLoad {
  init: LoadInit,
  state: LoadState,
  // ...
}

enum LoadInit {
  Main(String), // specifier
  MainWithCode(String, String), // specifier, code
  DynamicImport(String, String), // specifier, referrer
}

enum LoadState {
  Init,
  LoadingRoot,
  LoadingImports,
  Done,
}
```
It's a lot more intuitive and usable. Reading `impl Stream for RecursiveModuleLoad {}` is a lot easier.

Reply:
@nayeemrmn just to be clear, nothing changed in the actual behavior of `RecursiveModuleLoad`?

Reply:
> nothing changed in the actual behavior of `RecursiveModuleLoad`?

Correct, it's a refactor.